### PR TITLE
RATIS-1549. If the stream client is abnormally disconnected, reconnec…

### PR DESCRIPTION
**Pull request by guohao-rosicky**
## What changes were proposed in this pull request?

If the stream client is abnormally disconnected, reconnect the stream client.

Such as:
1. Socket channle is close
2. The network connection is down

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1549

### CODE REVIEW

**Comment by guohao-rosicky** [comment]
What do you think of this feature? @szetszwo

**Comment by guohao-rosicky** [comment]
1. Ctx.writeAndFlush() actually puts the request into the socket buffer. It just returns a Future object, and we need to add a listener to get its status
2. If channel inactive, we will fail to send with the current channel, so we need to trigger reconnection
3. When we connect, we add a listener to get the state of the connection

**Comment by szetszwo** [comment]
@guohao-rosicky , please add some tests.  We cannot easily tell if the change is working.

**Comment by guohao-rosicky** [comment]
Okay, I'll make a UT @szetszwo

**Comment by guohao-rosicky** [comment]
@captainzmc  can you  talk about the test results of this patch?

**Comment by captainzmc** [comment]
Thanks @guohao-rosicky  update this. I did a comparative test and found that this patch can indeed optimize the problem of client hung.  While testing three clients to write 400 files concurrently, ratis master hung every time, but RATIS-1549 runs normally.
image

Hi @szetszwo, In UT it is difficult to simulate a broken connection during write. Can we continue to discuss this patch based on the above test results?

**Comment by szetszwo** [comment]
@captainzmc , @guohao-rosicky , thanks for testing it manually.  I will review this change first.  Let me think about how to add a unit test.

**Comment by szetszwo** [comment]
@guohao-rosicky , I suggest to refactor the connection related code to a class; see https://issues.apache.org/jira/secure/attachment/13041023/620_review.patch

**Comment by guohao-rosicky** [comment]
@szetszwo  I have applied your patch

File: ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java [hunk 1]
```diff
@@ -125,30 +139,109 @@ int size() {
...
+      getWorkerGroup().schedule(this::reconnect, RECONNECT.getDuration(), RECONNECT.getUnit());
+    }
+
+    private ChannelFuture reconnect() {
+      final MemoizedSupplier<ChannelFuture> supplier = MemoizedSupplier.valueOf(this::connect);
+      final ChannelFuture previous = ref.getAndUpdate(prev -> prev == null? null: supplier.get());
+      if (previous != null) {
+        previous.channel().close();
+      }
+      return supplier.get();
```
Comment (szetszwo) [inline ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java:209]: We should check supplier.isInitialized().
```
      return supplier.isInitialized()? supplier.get(): null;
```

**Review by szetszwo — COMMENTED** [review]
@guohao-rosicky , thanks for the update.  Just two minor comments inlined.

File: ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java [hunk 2]
```diff
@@ -261,15 +362,27 @@ protected void decode(ChannelHandlerContext context, ByteBuf buf, List<Object> o
...
+      f.completeExceptionally(new AlreadyClosedException(this + ": Failed to send " + request));
+      return f;
+    }
     LOG.debug("{}: write {}", this, request);
-    getChannel().writeAndFlush(request);
+    channel.writeAndFlush(request).addListener(future -> {
+      if (!future.isSuccess()) {
+        final IOException e = new IOException(this + ": Failed to send " + request, future.cause());
+        LOG.error("Channel write failed", e);
+        if (!f.isDone()) {
```
[the remark this answers is not in the document]
Comment (guohao-rosicky) [reply ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java:375]: I see. Thank you

File: ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java (hunk 1 above)
Comment (guohao-rosicky) [reply ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java]: Has been modified. @szetszwo Please take a look.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**